rest/admin_api.go

```diff
@@ -818,3 +818,92 @@ func (h *handler) handlePurge() error {
 
 	return nil
 }
+
+// sg-replicate endpoints
+func (h *handler) getReplications() error {
+	replications, err := h.db.SGReplicateMgr.GetReplications()
+	if err != nil {
+		return err
+	}
+	bytes, err := base.JSONMarshal(replications)
+	h.writeRawJSON(bytes)
+	return err
+}
+
+func (h *handler) getReplication() error {
+	replicationID := mux.Vars(h.rq)["replicationID"]
+	replication, err := h.db.SGReplicateMgr.GetReplication(replicationID)
+	if replication == nil {
+		if err == nil {
+			return kNotFoundError
+		}
+		return err
+	}
+
+	bytes, err := base.JSONMarshal(replication)
+	h.writeRawJSON(bytes)
+	return err
+}
+
+func (h *handler) putReplication() error {
+	replicationID := mux.Vars(h.rq)["replicationID"]
+	return h.upsertReplication(replicationID)
+}
+
+func (h *handler) upsertReplication(replicationID string) error {
+	body, _ := h.readBody()
+	replicationConfig := &db.ReplicationCfg{}
+	if err := base.JSONUnmarshal(body, replicationConfig); err != nil {
+		return err
+	}
+
+	if h.rq.Method == "PUT" {
+		if replicationConfig.ID != "" && replicationConfig.ID != replicationID {
+			return base.HTTPErrorf(http.StatusBadRequest, "Replication ID in body %q does not match request URI", replicationConfig.ID)
+		}
+		replicationConfig.ID = replicationID
+	}
+
+	validateErr := replicationConfig.ValidateNewReplication(false)
+	if validateErr != nil {
+		return validateErr
+	}
+	created, err := h.db.SGReplicateMgr.UpsertReplication(replicationConfig)
+	if err != nil {
+		return err
+	}
+	if created {
+		h.writeStatus(http.StatusCreated, "Created")
+	}
+
+	return nil
+}
+
+func (h *handler) deleteReplication() error {
+	replicationID := mux.Vars(h.rq)["replicationID"]
+	return h.db.SGReplicateMgr.DeleteReplication(replicationID)
+}
+
+func (h *handler) getReplicationsStatus() error {
+	h.writeJSON(h.db.SGReplicateMgr.GetReplicationStatusAll())
+	return nil
+}
+
+func (h *handler) getReplicationStatus() error {
+	replicationID := mux.Vars(h.rq)["replicationID"]
+	h.writeJSON(h.db.SGReplicateMgr.GetReplicationStatus(replicationID))
+	return nil
+}
+
+func (h *handler) putReplicationStatus() error {
+	replicationID := mux.Vars(h.rq)["replicationID"]
+
+	action := h.getQuery("action")
+	if action == "" {
+		return base.HTTPErrorf(http.StatusBadRequest, "Query parameter 'action' must be specified")
+	}
+
+	return h.db.SGReplicateMgr.PutReplicationStatus(replicationID, action)
+}
+
+// Delete tests need to cover 'not found'
```

Comment:
Just so it gets highlighted in the editor clearly

```suggestion
// TODO: Delete tests need to cover 'not found'
```

Reply:
I should probably just write the DELETE tests now - I'll do that.